### DISCUSSION TITLE
Lesson 12: function depositCollateral | Clarification needed

### DISCUSSION THREAD

**Discussion by shrutic11** [Cyfrin/foundry-full-course-cu#980] [Q&A] [answered]
```solidity
 function depositCollateral(address tokenCollateralAddress, uint256 amountCollateral)
        public
        moreThanZero(amountCollateral)
        nonReentrant
        isAllowedToken(tokenCollateralAddress)
    {
        s_collateralDeposited[msg.sender][tokenCollateralAddress] += amountCollateral;
        emit CollateralDeposited(msg.sender, tokenCollateralAddress, amountCollateral);
        bool success = IERC20(tokenCollateralAddress).transferFrom(msg.sender, address(this), amountCollateral);
        if (!success) {
            revert DSCEngine__TransferFailed();
        }
    }
```

In this function, we are updating the mapping before performing the actual transfer :
```solidity
s_collateralDeposited[msg.sender][tokenCollateralAddress] += amountCollateral;
        emit CollateralDeposited(msg.sender, tokenCollateralAddress, amountCollateral);
        bool success = IERC20(tokenCollateralAddress).transferFrom(msg.sender, address(this), amountCollateral);
```

What happens if the transfer fails? Does the mapping still get updated?
In an ideal situation, shouldn't we first make the transfer, receive a positive confirmation, and then update the mapping?

**Comment by 0xusmanf** [comment, marked as answer]
> What happens if the transfer fails? Does the mapping still get updated?

In case of a failed transfer the transaction will revert (undoing all the changes made to the contract during that transaction) due to the following check. So no mapping will not get updated.
```solidity
if (!success) {
            revert DSCEngine__TransferFailed();
        }
```

> In an ideal situation, shouldn't we first make the transfer, receive a positive confirmation, and then update the mapping?

We are doing this to implement the [CEI](https://fravoll.github.io/solidity-patterns/checks_effects_interactions.html) -> Checks, Effects, Interactions method to avoid [reentrency attack](https://www.geeksforgeeks.org/reentrancy-attack-in-smart-contracts/).
